File: ils.R
#    Copyright: Matthijs den Besten
#
#    This file is part of Metascheduling.
#
#    Metascheduling is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    Metascheduling is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with Metascheduling.  If not, see <http://www.gnu.org/licenses/>.
#

# $Id: ils.R,v 1.21 2004/02/18 16:48:40 mldb Exp $ #

ils.race <- function(problem.id=sample(1:24, 1), maxExpScaling=10,
                     time.limit=60) {
  source("eRace.R", local=TRUE);
  scheduling.problems <- cbind(expand.grid(env=c("Fm", "Fm",
                                           "Pm", "Pm", "Pm", "S"),
                                           obj=c("T", "U"), weight=c(T, F)),
                               n=rep(c(50, rep(100, 5)), 4),
                               m=rep(c(20, 20, 2, 3, 4, 1), 4));

 
  vns.params <- function(idx) {
    vns.design <- expand.grid(P = factor(c("B", "F")),
                              A = factor(c("", "b", "f", "s",
                                "bf", "bs", "fs", "bsf")),
                              B = factor(c("", "b", "f", "s",
                                "bf", "bs", "fs", "bsf")),
                              C = factor(c("", "b", "f", "s",
                                "bf", "bs", "fs", "bsf")));
    attach(vns.design[idx,]);
    prefix <- paste("-L", P, sep="");
    i <- ifelse(A=="", "", paste(prefix, A, sep=""));
    ii <- ifelse(B=="", "", paste(prefix, B, sep=""));
    iii <- ifelse(C=="", "", paste(prefix, C, sep=""));
    detach(vns.design[idx,]);
    return(paste(i, ii, iii, collapse=""));
  }

  ls.cands <- sapply(c(419,182,406,1,405,453,307,168,167,181,279,280,451,420,40,566,789,790,54,454,676,22,21,229,231,535,565,675,35,8,53,23,24,39,661,85,679,694,791,917,919,118,950,819,308,452,72,58,6,30,57,103,285,339,51,548,71,595,613,693,741,918,935,230,232,343,533,536,551,597,599,86,662,855,936,947,998,7,70,116,69,120,123,10,5,99,126,75,101,117,16,3,29,100,932,104,108,236,413,534,563,614,619,87,709,742,792,982,983,186,340,804,314,36,68,803,838,172,564,612,43,171,235,344,552,555,616,680,854,856,920,939,997,310,4,25,67,102,106,44,107,14,9,76,296,119,15), vns.params);
  ils.candidates <- expand.grid(L = ls.cands,
                                P = "r",
                                l= 1, d = 9,
                                A = 0);
    
  make.task <- function(idx, specs) {
    source("problem.R", local=TRUE);
    source("dispatch.R", local=TRUE);
    inst <- tempfile("inst");
    seq <- tempfile("seq");
    
    this.tf <- runif(1);
    this.rdd <- runif(1);
    env <- specs$env;
    obj <- specs$obj;
    n <- specs$n;
    m <- specs$m;
    wght <- factor(ifelse(specs$weight, "unif", "common"),
                   levels=c("unif", "common"));

    ta.no <- sample(1:10,1) + ifelse(n==50, 50, 80);
    
    
    library(rpart);
    load("k.rp.Rbin");
    k <- 2^predict(prune(k.rp, 2^-8),
                   list(tf=this.tf,
                        rdd=this.rdd,
                        obj=obj,
                        w=wght,
                        n=n,
                        m=m,
                        env=env,
                        p=factor("unif")));
    if(env=="Fm") {
      instance <- read.instance.taillard(number = ta.no,
                                         TF = this.tf, RDD = this.rdd,
                                         wdis = wght,
                                         dir="../dat/");
      write.instance.Fm(instance, inst);
      
      cat(au.Fm(instance$p, instance$w, instance$d, k) -1, file=seq);
      
    } else {
      instance <- getInstance(njobs=n, TF=this.tf, RDD=this.rdd, wdis= wght,
                              mProcs=m);
      write.instance(instance, inst);
      
      if(env=="Pm") {
        cat(au.Pm(instance$p, instance$w, instance$d, m, k) -1, file=seq);
      } else {
        cat(au(instance$p, instance$w, instance$d, k) -1, file=seq);
      }
      
    }
    
    return(list(seq=seq, inst=inst,
                tf=this.tf, rdd=this.rdd, env=env, obj=obj, n=n, m=m, w=wght));
  }

  ils.string <- function(cand.spec, problem.spec, task.spec) {
    spec <- problem.spec;
    inst <- task.spec$inst;
    seq <- task.spec$seq;
    
    problem.str <- paste(ifelse(spec$env=="S", "-S",
                                ifelse(spec$env=="Fm", "-F", "-P")),
                         ifelse(spec$obj=="U", "-U", "-T"));
    attach(cand.spec);
    return(paste("cat", seq,
                 "| ./neh.out", problem.str, "-i", inst,
                 "| ./ils.out -R", problem.str, "-i", inst, 
                 "-t", time.limit, 
                 "-X", P, "-l", l, "-u", l+d,
                 "-A", A, L, "2>/dev/null"));
  }

  race.name <- paste("ils", problem.id, sep=".");
  
  invisible(race(wrapper.file="ils.race.wrapper.R",
                 maxExp=nrow(ils.candidates)*maxExpScaling,
                 log.file=paste(race.name, "race-log", sep="."),
                 problem=scheduling.problems[problem.id,],
                 candidates=ils.candidates,
                 task.fun=make.task,
                 cand.fun=ils.string,
                 race.name=race.name,
                 time.limit=time.limit,
                 ils.log=paste(race.name, "rtd-log", sep=".")));
}

File: ils.cc
#    Copyright: Matthijs den Besten
#
#    This file is part of Metascheduling.
#
#    Metascheduling is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    Metascheduling is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with Metascheduling.  If not, see <http://www.gnu.org/licenses/>.
#

/* $Id: ils.cc,v 1.12 2004/04/15 20:39:21 mldb Exp $ */
#include "Localsearch.h"
#include "Flow.h"
#include "Parallel.h"
#include "Run.h"
#include "Timer.h"
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#include <fstream>
#include <vector>

const class Localsearch* strtols(const char* params) {
  bool bestImpr = false;
  bool bwd = false;
  bool fwd = false;
  bool swp = false;
  
  int i = 0;
  bool eos = false;
  do {
    switch(params[i]) {
    case 'B':
      bestImpr = true;
      break;
    case 'F':
      bestImpr = false;
      break;
    case 'b':
      bwd = true;
      break;
    case 'f':
      fwd = true;
      break;
    case 's':
      swp = true;
      break;
    case ' ':
      break;
    default:
      eos = true;
      break;
    }
    i++;
  } while(!eos);
  
  return new Localsearch(bwd, fwd, swp, bestImpr);
}


int main(int argc, char* argv[]) {
  char* instance = NULL;
  char env = 'S';
  char obj = 'T';
  bool print_eval = false;
  bool print_trace = false;
  bool verbose = false;
  int optimum = 0;

  vector<const class Localsearch*> ls;
  char perturb = 'b';
  int lower = 0;
  int upper = 10;
  int backtrack = 0;
  int limit = 0;
  unsigned int seed = time(NULL);
  
  int opt;
  while((opt = getopt(argc, argv, "+i:t:s:L:X:l:u:A:SPFTUvRVo:")) != EOF) {
    extern char* optarg;
    switch(opt) {
    case 'i':
      instance = optarg;
      break;
    case 'S':
    case 'P':
    case 'F':
      env = (char) opt;
      break;
    case 'T':
    case 'U':
      obj = (char) opt;
      break;
    case 't':
      limit = atoi(optarg);
      break;
    case 's':
      seed = atoi(optarg);
      break;
    case 'L': {
      ls.push_back(strtols(optarg));
    } break;
    case 'X':
      perturb = optarg[0];
      break;
    case 'l':
      lower = atoi(optarg);
      break;
    case 'u':
      upper = atoi(optarg);
      break;
    case 'A':
      backtrack = atoi(optarg);
      break;
    case 'v':
      print_eval = true;
      break;
    case 'R':
      print_trace = true;
      break;
    case 'V':
      verbose = true;
      break;
    case 'o':
      optimum = atoi(optarg);
      break;
    default:
      break;
    }
  }
  
  srand(seed);
  
  ifstream ifs(instance);
  
  unsigned int sz = 0;
  vector<int> v;
  unsigned int job;
  while(cin >> job) {
    v.push_back(job);
    sz++;
  }
  unsigned int sequence[sz];
  copy(v.begin(), v.end(), sequence);

  class Solution* sol = NULL;
  class Solution* candidate = NULL;
  class Problem problem(ifs, env);
  switch(env) {
  case 'S':
    sol = new SRSol(sequence, sz, &problem, obj);
    candidate = new SRSol(sequence, sz, &problem, obj);
    break;
  case 'P':
    sol = new PSol(sequence, sz, &problem, obj);
    candidate = new PSol(sequence, sz, &problem, obj);
    break;
  case 'F':
    sol = new Flow(sequence, sz, &problem, obj);
    candidate = new Flow(sequence, sz, &problem, obj);
    break;
  }

  class Timer clock;
  int kicks = lower;
  int badSteps = 0;
  if(print_trace) {
    cout << "0" << "\t" << sol->eval();
    cout << endl;
  }
  if(verbose) {
    sol->print(cerr);
    cout << "0" << "\t" << sol->eval();
    cout << endl;
  }

  do {

    // Perturbation
    unsigned int sz = candidate->size();
    for(int k = 0; k < kicks; k++) {
      int a = rand() % sz;
      int b;
      do {
	b = rand() % sz;
      } while(b == a);
      
      unsigned int i = min(a, b);
      unsigned int j = max(a, b);
      switch(perturb) {
      case 'b':
	candidate->doBwd(i, j);
	break;
      case 'f':
	candidate->doFwd(i, j);
	break;
      case 's':
	candidate->doSwp(i, j);
	break;
      default: {
	  int mv = rand();
	  if(mv < RAND_MAX/3) {
	    candidate->doBwd(i, j);
	  } else if(mv < 2*RAND_MAX/3) {
	    candidate->doFwd(i, j);
	  } else {
	    candidate->doSwp(i, j);
	  }
	}
      }
    }

    // Local Search
    int start;
    do {
      start = candidate->eval();
      vector<const class Localsearch*>::iterator it = ls.begin();
      while(it != ls.end() && !clock.alarm(limit)) {
	(*it)->search(candidate);
	it++;
      }
    } while(candidate->eval() < start && !clock.alarm(limit));
    
    // Acceptance
    if(candidate->eval() >= sol->eval()) {
      kicks = (kicks < upper) ? kicks+1 : lower;
      badSteps++;
      if(badSteps > backtrack) {
	//candidate->clone(sol);
	delete candidate;
	sol->setSeq(sequence);
	switch(env) {
	case 'S':
	  candidate = new SRSol(sequence, sz, &problem, obj);
	  break;
	case 'P':
	  candidate = new PRSol(sequence, sz, &problem, obj);
	  break;
	case 'F':
	  candidate = new Flow(sequence, sz, &problem, obj);
	  break;
	}
	badSteps = 0;
      }
    } else {
      delete sol;
      candidate->setSeq(sequence);
      switch(env) {
      case 'S':
	sol = new SRSol(sequence, sz, &problem, obj);
	break;
      case 'P':
	sol = new PRSol(sequence, sz, &problem, obj);
	break;
      case 'F':
	sol = new Flow(sequence, sz, &problem, obj);
	break;
      }
      badSteps = 0;
      if(print_trace) {
	cout << clock.elapsed_time(Timer::VIRTUAL) << "\t" << sol->eval();
	cout << endl;
      }
      if(verbose) {
	sol->print(cerr);
	cout << clock.elapsed_time(Timer::VIRTUAL) << "\t" << sol->eval();
	cout << endl;
      }
    }
    
  } while(!clock.alarm(limit) && sol->eval() > optimum);

  if(!print_trace && !verbose) {
    if(print_eval) {
      cout << sol->eval() << endl;
    } else {
      sol->print();
    }
  }
  delete candidate;
  delete sol;

  return 0;
}

